File: Year-2022/LeetCode-July-Challenge/ConstructBinaryTreefromPreorderandInorderTraversal.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
class Solution {
    private int preIdx = 0;
    private TreeNode build(int[] preorder, int[] inorder, int start, int end){
        if(start > end) return null;
        TreeNode root = new TreeNode(preorder[preIdx]);
        preIdx++; int inIdx = 0;
        for(int i = start; i <= end; ++i){
            if(root.val == inorder[i]){
                inIdx = i;
                break;
            }
        }
        root.left = build(preorder, inorder, start, inIdx - 1);
        root.right = build(preorder, inorder, inIdx + 1, end);
        return root;
    }
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return build(preorder, inorder, 0, inorder.length - 1);
    }
    public void inorder(TreeNode root){
        if(root != null){
            inorder(root.left);
            System.out.print(root.val + " ");
            inorder(root.right);
        }
    }
}
public class ConstructBinaryTreefromPreorderandInorderTraversal {
    public static void main(String[] args) {
        int[] preorder = {3,9,20,15,7}, inorder = {9,3,15,20,7};
        Solution s = new Solution();
        TreeNode root = s.buildTree(preorder, inorder);
        s.inorder(root);
    }
}

File: Year-2022/LeetCode-July-Challenge/ConstructBinaryTreefromPreorderandInorderTraversal.js
function TreeNode(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
}

var buildTree = function(preorder, inorder) {
    p = i = 0
    build = function(stop) {
        if (inorder[i] != stop) {
            var root = new TreeNode(preorder[p++])
            root.left = build(root.val)
            i++
            root.right = build(stop)
            return root
        }
        return null
    }
    return build()
};
const inorderTraversal = (root) =>{
    if(root != null){
        inorderTraversal(root.left);
        console.log(root.val);
        inorderTraversal(root.right);
    }
}

let preorder = [1,2], inorder = [2,1];
let root = buildTree(preorder, inorder);
inorderTraversal(root);
